### PR TITLE
 Add notification setting validation API

### PR DESCRIPTION
**Pull request by wenchajun**
Signed-off-by: root <dehaocheng@yunify.com>

### What type of PR is this?

/kind feature

### Which issue(s) this PR fixes:

Fixes #
https://github.com/kubesphere/kubesphere/issues/4087
### Special notes for reviewers:
```
```

### Does this PR introduced a user-facing change?

```release-note
 Users can now verify if the notification settings are valid
```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by yuswift** [comment]
please fix the conflict

File: pkg/kapis/notification/v2beta2/handler.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+type Result struct {
+	Code    int    `json:"Status"`
+	Message string `json:"Message"`
+}
+
+func newHandler() handler {
+	return handler{}
+}
+
+const VerifyUrl = "http://notification-manager-svc.kubesphere-monitoring-system.svc:19093/api/v2/verify"
```
Comment (yuswift) [inline pkg/kapis/notification/v2beta2/handler.go:23]: the const should move to the first line

File: pkg/kapis/notification/v2beta2/handler.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+func (h handler) Verify(request *restful.Request, response *restful.Response) {
+
```
Comment (yuswift) [inline pkg/kapis/notification/v2beta2/handler.go:26]: no black line here

File: pkg/kapis/notification/v2beta2/handler.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	req, err := http.NewRequest("POST", VerifyUrl, request.Request.Body)
+	if err != nil {
+		response.WriteHeaderAndEntity(http.StatusInternalServerError, err)
+		return
+	}
+
+	req.Header.Set("Content-Type", "application/json")
```
Comment (yuswift) [inline pkg/kapis/notification/v2beta2/handler.go:33]: Is it really necessary? What's the default header?

File: pkg/kapis/notification/v2beta2/handler.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		response.WriteHeaderAndEntity(http.StatusInternalServerError, err)
+		return
+	}
+
+	defer resp.Body.Close()
```
Comment (yuswift) [inline pkg/kapis/notification/v2beta2/handler.go:42]: if the err != nil the defer func won't be called.

File: pkg/kapis/notification/v2beta2/handler.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	req, err := http.NewRequest("POST", VerifyUrl, request.Request.Body)
+	if err != nil {
+		response.WriteHeaderAndEntity(http.StatusInternalServerError, err)
+		return
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
```
Comment (yuswift) [inline pkg/kapis/notification/v2beta2/handler.go:36]: use defer here

File: pkg/kapis/notification/v2beta2/handler.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	if receiver.Labels["type"] == "tenant" {
+		if user != receiver.Labels["user"] {
+			response.WriteAsJson(Result{
+				http.StatusForbidden,
+				"No enough permissions",
+			})
+			return
+		}
+	}
+	if receiver.Labels["type"] == "default" || receiver.Labels["type"] == "global" {
```
Comment (wanjunlei) [inline pkg/kapis/notification/v2beta2/handler.go:60]: The receiver type is `tenant` or `global`, not `default`.

File: pkg/kapis/notification/v2beta2/handler.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	"io/ioutil"
+	"net/http"
+
+	"kubesphere.io/api/notification/v2beta2"
+
+	"github.com/emicklei/go-restful"
+)
+
+const (
+	VerifyUrl = "http://notification-manager-svc.kubesphere-monitoring-system.svc:19093/api/v2/verify"
```
Comment (wansir) [inline pkg/kapis/notification/v2beta2/handler.go:16]: Use a configurable variable instead is more appropriate. I think the host is not always `http://notification-manager-svc.kubesphere-monitoring-system.svc:19093`.

File: pkg/kapis/notification/v2beta2/register.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+	"kubesphere.io/kubesphere/pkg/api"
+	"kubesphere.io/kubesphere/pkg/apiserver/runtime"
+)
+
+var GroupVersion = schema.GroupVersion{Group: "notification.kubesphere.io", Version: "v2beta2"}
+
+func AddToContainer(container *restful.Container) error {
+	h := newHandler()
+	ws := runtime.NewWebService(GroupVersion)
+	ws.Route(ws.POST("/verify").
```
Comment (wansir) [inline pkg/kapis/notification/v2beta2/register.go:36]:
```suggestion
	ws.Route(ws.POST("/configs/{config}/verify").
```

File: pkg/kapis/notification/v2beta2/register.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+		Reads("").
+		To(h.Verify).
+		Returns(http.StatusOK, api.StatusOK, http.Response{}.Body)).
+		Doc("Provide validation for notification-manager information")
+	ws.Route(ws.POST("/users/{user}/verify").
```
Comment (wansir) [inline pkg/kapis/notification/v2beta2/register.go:41]: It seems unnecessary, whether the status code is 200 or 403, `v2beta2.Config` is able to create directly, isn't it?

**Review by wanjunlei — APPROVED** [review]

**Comment by zryfish** [comment]
@wenchajun Please go through 
https://github.com/kubesphere/community/blob/master/developer-guide/development/code-review-guide.md https://github.com/kubesphere/community/blob/master/developer-guide/development/coding-conventions.md
and write a proper release note, if this isn't any, just write `None`

File: pkg/kapis/notification/v2beta2/register.go (hunk 8 above)
Comment (wenchajun) [reply pkg/kapis/notification/v2beta2/register.go:38]: This is the global configuration interface, just to distinguish it from user

File: pkg/kapis/notification/v2beta2/handler.go [hunk 10]
```diff
@@ -3,20 +3,24 @@ package v2beta2
...
+	"github.com/emicklei/go-restful"
 	"kubesphere.io/api/notification/v2beta2"
 
-	"github.com/emicklei/go-restful"
+	nm "kubesphere.io/kubesphere/pkg/simple/client/notification"
 )
 
 const (
-	VerifyUrl = "http://notification-manager-svc.kubesphere-monitoring-system.svc:19093/api/v2/verify"
+	VerifyAPI = "/api/v2/verify"
```
Comment (benjaminhuo) [inline pkg/kapis/notification/v2beta2/handler.go:18]: VerifyAPI => VerificationAPI

File: pkg/kapis/notification/v2beta2/handler.go [hunk 11]
```diff
@@ -27,11 +31,21 @@ type notification struct {
...
+		option,
+	}
 }
 
 func (h handler) Verify(request *restful.Request, response *restful.Response) {
+	opt := h.option
+	if opt == nil || len(opt.Endpoint) == 0 {
+		response.WriteAsJson(Result{
+			http.StatusBadRequest,
+			"The corresponding Notification Manager path could not be found",
```
Comment (benjaminhuo) [inline pkg/kapis/notification/v2beta2/handler.go:45]: =>
Cannot find Notification Manager endpoint

File: pkg/kapis/notification/v2beta2/register.go (hunk 8 above)
Comment (benjaminhuo) [reply pkg/kapis/notification/v2beta2/register.go:38]: Can we change this to /configs/notification/verify ? @wansir @wenchajun @wanjunlei

File: pkg/kapis/notification/v2beta2/handler.go [hunk 12]
```diff
@@ -1,44 +1,85 @@
...
 
-	req.Header.Set("Content-Type", "application/json")
+	receiver := notification.Receiver
+	user := request.PathParameter("user")
+
+	if receiver.Labels["type"] == "tenant" {
+		if user != receiver.Labels["user"] {
+			response.WriteAsJson(Result{
+				http.StatusForbidden,
+				"No enough permissions",
```
Comment (benjaminhuo) [inline pkg/kapis/notification/v2beta2/handler.go:69]: Permission denied

File: pkg/kapis/notification/v2beta2/handler.go [hunk 13]
```diff
@@ -1,44 +1,85 @@
...
+			})
+			return
+		}
+	}
+	if receiver.Labels["type"] == "global" {
+		if user != "" {
+			response.WriteAsJson(Result{
+				http.StatusForbidden,
+				"No enough permissions",
```
Comment (benjaminhuo) [inline pkg/kapis/notification/v2beta2/handler.go:78]: Permission denied

**Comment by benjaminhuo** [comment]
Neither this PR's title nor the release notes is clear enough:

Add an API to provide forward validation => Add notification setting validation API
users can check whether the configuration is correct on the front end => Users can now verify if the notification settings are valid

**Pull request merged**